### PR TITLE
[Enhancement]lazy decode for parquet reader lazy columns

### PR DESCRIPTION
**Pull request by zombee0**
## What I'm doing:
ssbflat with high concurrency make cpu usage 100% and test the performance
  | new latency | old latency | improvement
-- | -- | -- | --
Q2.1 | 19.1 | 24.6 | 1.29
Q3.2 | 23.1 | 30.4 | 1.32
Q4.1 | 29.6 | 33.3 | 1.13
Q4.2 | 32.9 | 40.7 | 1.24
Q4.3 | 26.3 | 29.3 | 1.11

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by dirtysalt — APPROVED** [review]

File: be/src/formats/parquet/column_reader.cpp [hunk 1]
```diff
@@ -257,6 +263,14 @@ class ScalarColumnReader : public ColumnReader {
     const TypeDescriptor* _col_type = nullptr;
     const tparquet::ColumnChunk* _chunk_metadata = nullptr;
     std::unique_ptr<ColumnOffsetIndexCtx> _offset_index_ctx;
+
+    // _can_lazy_decode means string type and all page dict code
```
Comment (DorianZheng) [inline be/src/formats/parquet/column_reader.cpp:267]: It seems this optimization does not only works for string type but decimal, double types, etc too

**Review by Youngwb — APPROVED** [review]

File: be/src/formats/parquet/column_reader.cpp (hunk 1 above)
Comment (zombee0) [reply be/src/formats/parquet/column_reader.cpp:267]: yes, but we have not support support reading dict_code for decimal and double types, and we also don't support bytes->decimal convert for parquet, so up to now, we don't need to do this.

**Comment by eshishki** [comment]
i tested this commit with my sample production queries
```
with
curr_transactions as (
    select * from transactions
),
curr_acc_hist as (
    select * from accounts_history
),
curr_accounts as (
    select * from accounts
),
acc_hist_shortcut as (
    SELECT
        aa.tr_id,
        aa.operation_type,
        a.shortcut,
        aa.account_id
    FROM curr_acc_hist aa
    INNER JOIN curr_accounts a on aa.account_id = a.id

),
transactions as (
    select *,
        case when dest_acc = ah.account_id or dest_acc_name = ah.shortcut then 'income'
        else 'expense' end as transaction_type
    from curr_transactions tr join acc_hist_shortcut ah ON tr.id = ah.tr_id
    where status = 'approved'
    AND convert_tz(created_at,'Etc/UTC', 'Asia/Dushanbe')
    between '2023-02-01' and '2023-02-11'

)
select * from transactions limit 100;
``` 
gives 
```
ERROR 1064 (HY000): FileReader::get_next failed. reason = Internal error: DictDecoder append strings to column failed
be/src/formats/parquet/stored_column_reader.cpp:338 _read_values_on_levels(num_values, content_type, dst, false)
be/src/formats/parquet/complex_column_reader.cpp:198 _child_readers[field_name]->read_range(range, filter, child_column)
be/src/formats/parquet/group_reader.cpp:184 _column_readers[slot_id]->read_range(range, filter, (*chunk)->get_column_by_slot_id(slot_id))
be/src/formats/parquet/group_reader.cpp:1
```

i couldn't produce a self contained test case unfortunately
but this query works in 3.2.7 and fails on main branch with lazy decode

**Comment by zombee0** [comment]
@eshishki  could you check the be/log if there is more log message? and could you check this query just without this pr?

**Comment by eshishki** [comment]
@zombee0 i have verified that the issue persist in main branch and absent in branch 3.3 (without this pr), i've also verified that there is no issue up to this commit

**Comment by zombee0** [comment]
> @zombee0 i have verified that the issue persist in main branch and absent in branch 3.3 (without this pr), i've also verified that there is no issue up to this commit

will be fixed by https://github.com/StarRocks/starrocks/pull/47147, thank you!

**Pull request merged**